src/ocaml/mode.ml

```diff
@@ -135,3 +137,135 @@ module Dict = struct
+    | Only m -> Variant ("Only", [ to_dyn m ])
+
+  let encode t =
+    let open Dune_sexp.Encoder in
+    match t with
+    | All -> string "all"
+    | Only m -> encode m
+
+  let decode =
+    let open Dune_sexp.Decoder in
+    let+ value = either (keyword "all") decode in
+    match value with
+    | Left () -> All
+    | Right mode -> Only mode
+end
+
+module MultiDict = struct
+  (* Here we use a Hashtbl to be able in the future to use more variants than
+     just "Byte, Native or All". Since values are always added and never deleted
+     or modified we take advantage of the fact that when using [add t key value]
+     previous bindings for [key] are not removed but simply hidden and that
+     [find_all] returns all these bindings to represent list of values with a
+     hashtbl for single values. *)
+  module Hashtbl = Hashtbl.Make (Select)
+
+  type mode = t
+
+  type 'a t = 'a Hashtbl.t
+
+  let create () = Hashtbl.create 30
+
+  let add_multiple t for_mode l =
+    List.iter ~f:(Hashtbl.add_over t for_mode) l;
+    t
+
+  let create_for_all_modes l =
+    let tbl = create () in
+    add_multiple tbl All l
+
+  let for_all_modes t = Hashtbl.find_all t All
+
+  let for_only ?(and_all = false) t mode =
+    let all = if and_all then for_all_modes t else [] in
+    Hashtbl.find_all t (Only mode) |> List.rev_append all
+
+  let all t =
+    Hashtbl.to_seq_values t
+    |> Seq.fold_left ~init:[] ~f:(fun acc value -> value :: acc)
+
+  let add t for_mode v =
+    Hashtbl.add_over t for_mode v;
+    t
+
+  let encode encoder t =
+    let open Dune_sexp.Encoder in
+    let fields =
+      Hashtbl.foldi t ~init:[] ~f:(fun for_ _file acc ->
+          let files = Hashtbl.find_all t for_ in
+          if List.is_empty files then acc
+          else
+            let field_for = field "for" Select.encode for_ in
+            let field_files = field_l "files" encoder files in
+            field_l "archives" Fun.id (record_fields [ field_for; field_files ])
+            :: acc)
+    in
+    record_fields fields
+
+  let decode decoder =
+    let open Dune_sexp.Decoder in
+    let+ fields =
+      fields
+      @@ multi_field "archives"
+           (fields
+              (let+ for_ = field "for" Select.decode
+               and+ files = field "files" (repeat decoder) in
+               (for_, files)))
+    in
+    let tbl = create () in
+    List.iter fields ~f:(fun (for_, paths) ->
+        ignore @@ add_multiple tbl for_ paths);
+    tbl
+
+  let to_dyn to_dyn t =
+    let open Dyn in
+    let l =
+      Hashtbl.foldi t ~init:[] ~f:(fun for_ _file acc ->
+          let files = Hashtbl.find_all t for_ in
+          Record
+            [ ("for_mode", Select.to_dyn for_); ("files", list to_dyn files) ]
+          :: acc)
+    in
+    List l
+
+  let equal eq t t' =
+    Seq.equal
+      (fun (f1, p1) (f2, p2) -> Select.equal f1 f2 && eq p1 p2)
+      (Hashtbl.to_seq t) (Hashtbl.to_seq t')
+
+  let remapi ~f t =
+    let new_t = Hashtbl.create 30 in
```

Comment:
this should be either `create ()` or use the size of `t`

Reply:
obsolete